tools/pybind11Tools.cmake

```diff
@@ -37,25 +37,74 @@ function(select_cxx_standard)
 endfunction()
 
 # Internal: find the appropriate LTO flag for this compiler
-macro(_pybind11_find_lto_flag output_var prefer_thin_lto)
-  if(${prefer_thin_lto})
-    # Check for ThinLTO support (Clang)
-    check_cxx_compiler_flag("-flto=thin" HAS_THIN_LTO_FLAG)
-    set(${output_var} $<${HAS_THIN_LTO_FLAG}:-flto=thin>)
+function(_pybind11_find_lto_flag cxxflags_out linkerflags_out prefer_thin_lto)
+  # Release type (upper case for simpler comparisons)
+  string(TOUPPER "${CMAKE_BUILD_TYPE}" U_CMAKE_BUILD_TYPE)
+
+  set(try_cxx_flags "")
+  set(try_linker_flags "")
+
+  if(U_CMAKE_BUILD_TYPE MATCHES REL AND CMAKE_CXX_COMPILER_ID MATCHES "^(GNU|Clang)$")
+    # Enable link time optimization
+    if(NOT CMAKE_CXX_FLAGS MATCHES "-flto")
+      set(cxx_append "")
+      set(linker_append "")
+      if (CMAKE_CXX_COMPILER_ID MATCHES "Clang" AND NOT APPLE AND U_CMAKE_BUILD_TYPE MATCHES "MINSIZEREL")
+        # Clang Gold plugin does not support -Os
+        set(linker_append " -O3")
+      elseif(CMAKE_CXX_COMPILER_ID MATCHES "GNU")
+        set(cxx_append " -fno-fat-lto-objects")
+      endif()
+
+      if (CMAKE_CXX_COMPILER_ID MATCHES "Clang" AND ${prefer_thin_lto})
+        list(APPEND try_cxx_flags "-flto=thin${cxx_append}")
+        list(APPEND try_linker_flags "-flto=thin${linker_append}")
+      endif()
+
+      list(APPEND try_cxx_flags "-flto${cxx_append}")
+      list(APPEND try_linker_flags "-flto${linker_append}")
+    endif()
+  elseif (CMAKE_CXX_COMPILER_ID MATCHES "Intel")
+    # Intel equivalent to LTO is called IPO
+    list(APPEND try_cxx_flags "-ipo")
+    list(APPEND try_linker_flags "-ipo")
+  elseif(MSVC)
+    list(APPEND try_cxx_flags "/GL")
+    # cmake only interprets libraries as linker flags when they start with a - (otherwise it
+    # converts /LTCG to \LTCG as if it was a Windows path).  Luckily MSVC supports passing arguments
+    # with - instead of /, even if it is a bit non-standard:
+    list(APPEND try_linker_flags "-LTCG")
   endif()
 
-  if(NOT ${prefer_thin_lto} OR NOT HAS_THIN_LTO_FLAG)
-    if(NOT CMAKE_CXX_COMPILER_ID MATCHES "Intel")
-      # Check for Link Time Optimization support (GCC/Clang)
-      check_cxx_compiler_flag("-flto" HAS_LTO_FLAG)
-      set(${output_var} $<${HAS_LTO_FLAG}:-flto>)
+  set(CMAKE_REQUIRED_QUIET 1)
+  set(orig_CMAKE_REQUIRED_LIBRARIES ${CMAKE_REQUIRED_LIBRARIES})
+  list(LENGTH try_cxx_flags try_max)
+  math(EXPR try_max ${try_max}-1)
```

Comment:
When `try_cxx_flags` is empty (as it is on OSX due to the failed check above), this ends up setting `try_max` to -1. That, unfortunately, causes the `RANGE` to apparently try to execute...forever, which causes a cmake error when doing the `list(GET)`.